src/utilities/BatchTableHierarchyExtension.js

```diff
@@ -0,0 +1,123 @@
+import { parseBinArray } from './FeatureTable.js';
+
+export class BatchTableHierarchyExtension {
+
+	constructor( batchTable ) {
+
+		this.batchTable = batchTable;
+
+		const extensionHeader = batchTable.header.extensions[ '3DTILES_batch_table_hierarchy' ];
+
+		this.classes = extensionHeader.classes;
+		for ( const classDef of this.classes ) {
+
+			const instances = classDef.instances;
+			for ( const property in instances ) {
+
+				classDef.instances[ property ] = this._parseProperty( instances[ property ], classDef.length );
+
+			}
+
+		}
+
+		this.instancesLength = extensionHeader.instancesLength;
+
+		this.classIds = this._parseProperty( extensionHeader.classIds, this.instancesLength );
+		this.parentCounts = this._parseProperty( extensionHeader.parentCounts, this.instancesLength )
+			|| new Array( this.instancesLength ).fill( 1 );
+		this.parentIds = this._parseProperty( extensionHeader.parentIds );
+
+		this.instancesIds = [];
+		const classCounter = {};
+		for ( const classId of this.classIds ) {
+
+			classCounter[ classId ] = classCounter[ classId ] ?? 0;
+			this.instancesIds.push( classCounter[ classId ] );
+			classCounter[ classId ] ++;
+
+		}
+
+	}
+
+	_parseProperty( property, propertyLength ) {
+
+		if ( ! ( property instanceof Object ) ) {
+
+			return property;
+
+		} else if ( Array.isArray( property ) ) {
+
+			return property;
+
+		} else {
+
+			const { buffer, binOffset } = this.batchTable;
+
+			const byteOffset = property.byteOffset;
+			const componentType = property.componentType || 'UNSIGNED_SHORT';
+
+			const arrayStart = binOffset + byteOffset;
+
+			return parseBinArray( buffer, arrayStart, propertyLength, 'SCALAR', componentType );
+
+		}
+
+	}
+
+	getDataFromId( id ) {
+
+		const properties = {};
+
+		// Get properties inherited from parents
+
+		const parentCount = this.parentCounts[ id ];
+
+		if ( this.parentIds && parentCount > 0 ) {
+
+			let parentIdsOffset = 0;
+			for ( let i = 0; i < id; i ++ ) {
+
+				parentIdsOffset += this.parentCounts[ i ];
+
+			}
+
+			const parentIds = this.parentIds.slice(
+				parentIdsOffset,
+				parentIdsOffset + parentCount,
+			);
+
+			for ( const parentId of parentIds ) {
```

Comment:
Can we just use a normal for loop here instead of making a new array?

```js
for ( let i = 0; i < parentCount; i ++ ) {

  const parentId = parentIds[ parentIdsOffset + i ];
```

Reply:
Indeed, it is simpler. I changed it.